flang/lib/Lower/CallInterface.cpp

```diff
@@ -0,0 +1,455 @@
+         passedEntity.passBy != CallInterface::PassEntityBy::AddressAndLength &&
+         "bad arg position");
+  actualInputs[passedEntity.firArgument] = arg;
+}
+
+void Fortran::lower::CallerInterface::placeAddressAndLengthInput(
+    const PassedEntity &passedEntity, mlir::Value addr, mlir::Value len) {
+  assert(static_cast<int>(actualInputs.size()) > passedEntity.firArgument &&
+         static_cast<int>(actualInputs.size()) > passedEntity.firLength &&
+         passedEntity.firArgument >= 0 && passedEntity.firLength >= 0 &&
+         passedEntity.passBy == CallInterface::PassEntityBy::AddressAndLength &&
+         "bad arg position");
+  actualInputs[passedEntity.firArgument] = addr;
+  actualInputs[passedEntity.firLength] = len;
+}
+
+bool Fortran::lower::CallerInterface::verifyActualInputs() const {
+  if (getNumFIRArguments() != actualInputs.size())
+    return false;
+  for (auto arg : actualInputs) {
+    if (!arg)
+      return false;
+  }
+  return true;
+}
+
+template <typename T>
+static inline auto AsGenericExpr(T e) {
+  return Fortran::evaluate::AsGenericExpr(Fortran::common::Clone(e));
+}
+
+mlir::Value Fortran::lower::CallerInterface::getResultLength() {
+  // FIXME: technically, this is a specification expression,
+  // so it should be evaluated on entry of the region we are
+  // in, it can go wrong if the specification expression
+  // uses a symbol that may have change.
+  //
+  // The characteristic has to be explicit for such
+  // cases, so the allocation could also be handled on callee side, in such
+  // case. For now, protect with an unreachable.
+  assert(characteristic && "characteristic was not computed");
+  const auto *typeAndShape =
+      characteristic->functionResult.value().GetTypeAndShape();
+  assert(typeAndShape && "no result type");
+  auto expr = AsGenericExpr(typeAndShape->LEN().value());
+  if (Fortran::evaluate::IsConstantExpr(expr))
+    return converter.genExprValue(expr);
+  llvm_unreachable(
+      "non constant result length on caller side not yet safely handled");
+}
+
+//===----------------------------------------------------------------------===//
+// Callee side interface implementation
+//===----------------------------------------------------------------------===//
+
+bool Fortran::lower::CalleeInterface::hasAlternateReturns() const {
+  return !funit.isMainProgram() &&
+         Fortran::semantics::HasAlternateReturns(funit.getSubprogramSymbol());
+}
+
+std::string Fortran::lower::CalleeInterface::getMangledName() const {
+  return funit.isMainProgram()
+             ? fir::NameUniquer::doProgramEntry().str()
+             : converter.mangleName(funit.getSubprogramSymbol());
+}
+
+mlir::Location Fortran::lower::CalleeInterface::getCalleeLocation() const {
+  // FIXME: do NOT use unknown for the anonymous PROGRAM case. We probably
+  // should just stash the location in the funit regardless.
+  return converter.genLocation(funit.getStartingSourceLoc());
```

Comment:
Yeah. We need a policy for that case. Use the location of the first statement?

Reply:
That sounds good to me, I 'll do a small PR with this (I guess it should be done inside getStartingSourceLoc PFT helper).